specs/networking/libp2p-standardization.md

```diff
@@ -0,0 +1,131 @@
+
+## Gossipsub
+
+#### Protocol id: `/meshsub/1.0.0`
+
+*Note: Parameters listed here are subject to a large-scale network feasibility
+study*
+
+The [Gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub)
+protocol will be used for block and attestation propagation across the
+network.
+
+### Configuration Parameters
+
+Gossipsub has a number of internal configuration parameters which directly
+effect the network performance.  Clients can implement independently, however
+we aim to standardize these across clients to optimize the gossip network for
+propagation times and message duplication. Current network-related defaults are:
+
+```
+(
+	// The target number of peers in the overlay mesh network (D in the libp2p specs).
+	mesh_size: 6
+	// The minimum number of peers in the mesh network before adding more (D_lo in the libp2p specs).
+	mesh_lo: 4
+	// The maximum number of peers in the mesh network before removing some (D_high in the libp2p sepcs).
+	mesh_high: 12
+	// The number of peers to gossip to during a heartbeat (D_lazy in the libp2p sepcs).
+	gossip_lazy: 6 // defaults to `mesh_size`
+	// Time to live for fanout peers (seconds).
+	fanout_ttl: 60
+	// The number of heartbeats to gossip about.
+	gossip_history: 3
+	// Time between each heartbeat (seconds).
+	heartbeat_interval: 1
+)
+```
+
+### Topics
+
+*The Go and Js implementations use string topics - This is likely to be
+updated to topic hashes in later versions - https://github.com/libp2p/rust-libp2p/issues/473*
+
+For Eth2.0 clients, topics will be sent as `SHA2-256` hashes of the topic string.
+
+There is one dedicated topic for propagating beacon blocks and aggregated
+attestations across the network. This topic will have the string
+`beacon_chain`. Each shard will have it's own topic allowing relevant parties
+to subscribe to in order to receive local shard attestations. The shard topics are
+prefixed with `shard` followed by the number of the shard. For example,
+messages relating to shard 10, will have the topic string `shard10`.
+
+### Messages
+
+Messages sent across gossipsub are fixed-size length-prefixed byte arrays.
+Each message has a maximum size of 512KB (estimated from expected largest uncompressed
+block size).
+
+The byte array is prefixed with a unsigned 64 bit length number encoded as an
+`unsigned varint` (https://github.com/multiformats/unsigned-varint). Gossipsub messages therefore take the form:
+```
++--------------------------+
+|     message length       |
++--------------------------+
+|                          |
+|       body (<1M)         |
+|                          |
++--------------------------+
+```
+
+The body of the message is an SSZ-encoded object representing either a
+beacon block or attestation. The type of objected is determined via a prefixed
+nibble. Currently there are two objects that are sent across the gossip
+network. They are (with their corresponding nibble specification):
+
+- `0x1`: Beacon block
```

Comment:
Not necessarily suggesting it, but if we use a different topic for beacon blocks and attestations, we would not need the nibble, correct?

The major downside I see in fragmenting topics is that it gives nodes more optionality on what to listen for which could reduce the robustness/density of any given topic.

Reply:
Yep, this is what @prestonvanloon was suggesting. I think multiple topics is a better approach assuming we only have two objects (and therefore topics).  I'll modify this.

Reply:
+1